translator/trace/jaeger/protospan_to_jaegerthrift.go

```diff
@@ -30,6 +30,8 @@ import (
 
 var (
 	unknownProcess = &jaeger.Process{ServiceName: "unknown-service-name"}
+	errorTag       = "error"
+	errorMessage   = "errorMessage"
```

Comment:
Where does this tag name come from? Is it documented for Jaeger?

Reply:
This tag is used by jaeger ui. Refer 

> func Sanitize

 at line number 167 at https://github.com/jaegertracing/jaeger/blob/master/cmd/collector/app/sanitizer/zipkin/span_sanitizer.go

Reply:
Perhaps a link to Jaeger spec (which I think is just OpenTracing) would be more helpful? OpenTracing does specify error tag as  a boolean but does not mention errorMessage AFAICT. https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table and https://github.com/opentracing/specification/blob/master/semantic_conventions.md#captured-errors

Reply:
Without this Zipkin to Jaeger Span translation it will break span filter error=true tag

Reply:
Looks like it uses `error.message`, not `errorMessage` https://github.com/jaegertracing/jaeger/blob/master/cmd/collector/app/sanitizer/zipkin/span_sanitizer.go#L180

Reply:
I think it is acceptable if this is a documented Jaeger convention even if not entirely covered by OpenTracing. However, this will break backends that depend on the current translation. So if we decide to accept this, it'll have to wait until the change is announced and people have a chance to update their error processing code.

Reply:
i have tested these changes with jaeger and it working fine. Which other backend depends on this translation ? The fix is targeting only jaeger spans, and moreover we are not removing any tags we are just changing the type and adding additional error.message tag.

Reply:
> Looks like it uses `error.message`, not `errorMessage`? https://github.com/jaegertracing/jaeger/blob/master/cmd/collector/app/sanitizer/zipkin/span_sanitizer.go#L180

sure, we can stick to error.message.

Reply:
> Perhaps a link to Jaeger spec (which I think is just OpenTracing) would be more helpful? OpenTracing does specify error tag as a boolean but does not mention errorMessage AFAICT. https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table and https://github.com/opentracing/specification/blob/master/semantic_conventions.md#captured-errors

Since Zipkin sets "error" tag as a string, converting string to boolean will lead to a lossy translation so additional "error.message" tag will help carry error message that was originally part of zipkin tag

Reply:
>Which other backend depends on this translation ? The fix is targeting only jaeger spans, and moreover we are not removing any tags we are just changing the type and adding additional error.message tag.

Translation from OC to Jaeger is a public code that is Exported by this repo. Changing its logic may break code that imports it and exists elsewhere. Such code definitely exists, here are just 2 examples that use jaegertranslator:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/kinesisexporter/exporter.go
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/sapmexporter/exporter.go

This PR has much wider effect than it appears on the surface.

The other problem with this PR is that it does affects attribute named "error" regardless of where they come from. It may be that "error" has the specific meaning in Zipkin, but this PR does not attempt to solve that particular case and instead has a much wider than necessary effect.

The proper way to implement translations is the following:
1. Identify the particular element of input format that needs to be translated.
2. In the corresponding receiver implement non-ambiguous translation from the input format to OpenCensus format, such that this translation is universal and does not imply what exporter will be used later. Sometimes this requires coming up with new semantic conventions since OpenCensus may not have the corresponding direct representation. In such cases first a PR must be submitted against OpenTelemetry specification repo to add the semantic convention (this step appears to be necessary in this case because "error" is not a native field in OpenCensus representation and does not exist in semantic conventions).
3. Identify how the representation of the element in OpenCensus format maps to the output format.
4. In the corresponding exporter implement non-ambiguous translation from OpenCensus format to the output format.
5. Add tests that verify behavior for all effected receivers and exporters, preferably as N*M matrix test.

Translations from input format to OpenCensus and from OpenCensus to output formats must be implemented independently, such that any combination of receivers and exporters works correctly. Collector must correctly perform data mapping from any receiver to any exporter.
 
Let me know if further clarification is needed. Until the translation is performed correctly I will have to block this PR.

Reply:
Instead of a change in translator, we could simply fix this I. ZipkinReceiver to localization the change. Your thoughts on this?

Reply:
@tigrannajaryan https://github.com/open-telemetry/oteps/pull/86

Reply:
>  Which other backend depends on this translation ? The fix is targeting only jaeger spans, and moreover we are not removing any tags we are just changing the type and adding additional error.message tag.

There are SaaS backends that rely on OpenTelemetry and could be affected. Plus, as authors of an open-source project, we don't have a list of projects/people using the software we ship. It's best to assume that making breaking changes will break systems we don't know about and I do know at least a couple of SaaS backends that reply on Otel exported Jaeger data.

Reply:
To clarify, I agree that this is a bug in the translator and should be fixed but I think it needs to be announced widely (may be in a coulpe of community meetings) and the community need to be given enough time to prepare for it.

Reply:
> > Which other backend depends on this translation ? The fix is targeting only jaeger spans, and moreover we are not removing any tags we are just changing the type and adding additional error.message tag.
> 
> There are SaaS backends that rely on OpenTelemetry and could be affected. Plus, as authors of an open-source project, we don't have a list of projects/people using the software we ship. It's best to assume that making breaking changes will break systems we don't know about and I do know at least a couple of SaaS backends that reply on Otel exported Jaeger data.

I agree but this is a bug breaking jaeger functionality, and it should be the case with other SaaS offerings accepting jaeger spans

Reply:
@owais can you clarify what exactly is the bug here? Is it in Zipkin to OC translation to in OC to Jaeger translation?

To make it clear what I am aiming for: the translation to and from OC needs to be protocol-neutral.

Neutrality means 2 things:
1. The attribute is not special and is simply passed through as-is, without interfering with it.
2. The attribute is special, has a special meaning in the particular protocol _and_ has a corresponding special in-memory representation in Collector. In this case a translation is needed. An example is translation of in-memory SpanKind field to Jaeger's "span.kind" tag (which has a special meaning and is defined in OpenTracing semantic conventions). 

For any other custom transformations processors should be used.

Question: does Zipkin define the "error" tag in any special way? I have not been able to find this in the docs. If it does can you post a link?

Reply:
> we could simply fix this in ZipkinReceiver to localize the change to zipkin alone but again zipkin to opencensus might have an issue.

@vikrambe this would be the right approach if the meaning of the "error" tag was defined in Zipkin specification. I cannot find  any references to this though. Do you have a link?

Reply:
@tigrannajaryan https://github.com/openzipkin/zipkin4net/issues/212

Reply:
@tigrannajaryan you're right. I totally overlooked that we are translating just tags here while as OC has a status field to represent error. If any formats use tags to represent errors, they should be translated to OC's status and then it'll automatically get translated to Jaeger later. If OC status to Jaeger error attribute or zipkin error (however it is represented) to OC doesn't work, we should fix that instead.

Reply:
@owais @tigrannajaryan Can i update the PR with OC.Status and get rid of error and error.message from opentracing spec ? please confirm ?

Reply:
@vikrambe let's see what Zipkin maintainer replies.

Also, it is not clear where the error message string should be stored when translating to Jaeger. This PR initially proposed "errorMessage", then "error.message". What does Jaeger spec say?

As explained above we cannot rush this change. It will be only accepted after full investigation and after ensuring we will not break other people's code that relies on current behavior. If you would like to expedite this more evidence in the form of authoritative links to specs and confirmation from other parties can help.

Reply:
@tigrannajaryan Sorry for the confusion, jaeger uses 'error.message' to carry the actual error message. I will correct it in the PR. 

Sure, i will see if i get more on zipkin spec. I went through the zipkin span model, there is no special mention about error tag. I will dig in to the zipkin code and try finding out.

Sure, i am not in hurry.  I wanted to fix this to help others facing same issue.

Reply:
Sounds good. Thanks for your help. Let's get this fixed correctly.

Reply:
Sure. Here is spring-cloud-sleuth documentation explaining error tag with the visualization
https://cloud.spring.io/spring-cloud-sleuth/reference/html/#visualizing-errors